transport/udp/udp.go

```diff
@@ -0,0 +1,390 @@
+// Copyright 2016 The Mangos Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use file except in compliance with the License.
+// You may obtain a copy of the license at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package udp implements the UDP transport for mangos. It is based on
+// M. Sustrik, Ed., "UDP Mapping for Scalability Protocols",
+// sp-udp-mapping-01 (March 2014).
+package udp
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/go-mangos/mangos"
+)
+
+// Errors produced by this transport prototocol
+var (
+	ErrMsgTruncated = errors.New("message truncated due to insufficient buffer space")
+)
+
+// connHeader is exchanged during the initial handshake.
+// FIXME: stolen from conn.go -- unify this
+type connHeader struct {
+	Zero    byte // must be zero
+	S       byte // 'S'
+	P       byte // 'P'
+	Version byte // only zero at present
+	Proto   uint16
+	Rsvd    uint16 // always zero at present
+}
+
+const (
+	// Maximum read queue depth for listening endpoints
+	// FIXME: instead of udpMaxBacklog, use a ListenerOption to set this
+	// FIXME: e.g. OptionUdpListenerBacklog
+	udpMaxBacklog = 100
+
+	// Maximum expected UDP datagram size.
+	// UDPv4 datagrams can be up to 65507 = 2^16-1 - 20 - 8 bytes, due to the 2-byte
+	// IPv4 total length field, a minimum IPv4 header size of 20 bytes, and the 8
+	// byte UDP header length. IPv6 minimum header size is 40 bytes, so it is slightly
+	// less (not considering UDPv6 jumbograms, RFC 2675).
+	// These are theoretical limits, since in practice the maximum datagram size
+	// is determined by the MTU (typically 1500 .. ~9000 bytes).
+	// Hence a buffer size of 64k should suffice for 'regular' networks.
+	udpMaxBuf = 1 << 16
```

Comment:
Reading the RFC, it seems that this is restricted only to specific links; so for most people it won't be available.  But one can imagine e.g. shared memory based networks (RDMA, etc.) where the link MTU can be quite large.  For example, we had a shared memory network in the Sun E10K between domains that made use of shared memory between the two domains.  That could have used a very large MTU, say 1MB, without concerns about checksum etc. because it was simply local memory to both nodes.  One can imagine a similar situation with paravirualized networks, although I believe today they all have "smaller" jumbo MTUs.